content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,230 @@ Like a segment tree on values rather than indices.
 	</Resource>
 </Resources>
 
-### Solution - Range K-th Smallest
+Suppose you want to support the following queries:
+
+- Given a range $l$, $r$ count the number of occurrences of value $x$.
+- Given a range $l$, $r$ find the $k$ smallest element
+
+With a wavelet tree, you can easily support those queries in $log(M)$ time.
+when M is the maximum value in the array.
+
+## Wavelet Tree Structure
+
+A wavelet tree is a binary tree where each node represents a range of values.
+The root node covers the entire range, and each subsequent level splits the
+range into two halves.
+
+We are going to maintain a segment tree over values instead of indices. Each segment will contain indices whose values lie within the segment's range. We'll save those indices in a vector. Notice that an index can be in at most $log(M)$ segments
+
+<Spoiler title="Wavelet tree Visualization">
+
+Let's say our array is: $[3,5,3,1,2,2,3,4,5,5]$
+
+![Wavelet Tree Visualization](./assets/diagram.png)
```

Comment:
ok i have no idea what each node in the diagram is supposed to mean lol

Reply:
Each node has an array representing the indecies of every number between l and r

Reply:
tell that to the reader, not me